### PR TITLE
[Inductor][CPP] Fix Half data type cse cache issue for CPP Backend

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128498

**Summary**
Fixing issue: https://github.com/pytorch/pytorch/issues/128263. After https://github.com/pytorch/pytorch/issues/115260, we cached the higher precision cse variable to avoid duplicate casting between buffers. However, it failed to check the original data type. This means if we convert `int32` to `bf16` for `store` and then convert `bf16` back to `fp32` for `load`, it would incorrectly hit the cache and reuse the `int32` cse var. This PR fixes the issue.

**Test Plan**
```
python -u -m pytest -s -v test/inductor/test_cpu_repro.py -k test_issue_128263
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Review by zhuhaozhe — APPROVED** [review]

**Comment by leslie-fang-intel** [comment]
Hi @jerryzh168, please help to take a look.

File: test/inductor/test_cpu_repro.py [hunk 1]
```diff
@@ -3657,6 +3657,30 @@ def fn(x):
         self.common(fn, (x,))
         assert metrics.generated_cpp_vec_kernel_count == 1
 
+    def test_issue_128263(self):
```
Comment (jerryzh168) [inline test/inductor/test_cpu_repro.py:3660]: maybe you want to change this to a more descriptive name?

**Review by jerryzh168 — APPROVED** [review]

File: test/inductor/test_cpu_repro.py (hunk 1 above)
Comment (leslie-fang-intel) [reply test/inductor/test_cpu_repro.py]: Thanks, changed to a more descriptive name.

**Pull request closed without merging**